### PR TITLE
proxy_filter: Fix the CONNECT implementation when the hostname contains a port number

### PR DESCRIPTION
**Pull request by fredyw**
Commit Message: Fixes the `CONNECT` implementation when the hostname contains a port number causing the `CONNECT` request created to be invalid.
Additional Description: When the port number is not specified, the port 443 will be automatically added just like before.
Risk Level: low
Testing: unit and integration tests
Docs Changes: inline
Release Notes: inline
Platform Specific Features: n/a
Optional Runtime guard: `envoy.reloadable_features.proxy_ssl_port`

### CODE REVIEW

File: changelogs/current.yaml [hunk 1]
```diff
@@ -128,6 +128,11 @@ bug_fixes:
     the number of requests per I/O cycle is configured and an HTTP decoder filter that pauses filter chain is present. This behavior
     can be reverted by setting the runtime guard ``envoy.reloadable_features.use_filter_manager_state_for_downstream_end_stream``
     to false.
+- area: http_proxy
```
Comment (alyssawilk) [inline changelogs/current.yaml:131]: clarify this is with the proxy filter (which is arguably one folks can't use as there's no way to configure it outside of E-M)

File: source/extensions/transport_sockets/http_11_proxy/connect.cc [hunk 2]
```diff
@@ -37,11 +37,17 @@ UpstreamHttp11ConnectSocket::UpstreamHttp11ConnectSocket(
   // options, we want to maintain the original behavior of this transport socket.
   if (options_ && options_->http11ProxyInfo()) {
     if (transport_socket_->ssl()) {
-      header_buffer_.add(
-          absl::StrCat("CONNECT ", options_->http11ProxyInfo()->hostname, ":443 HTTP/1.1\r\n\r\n"));
+      if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.proxy_ssl_port")) {
+        header_buffer_.add(absl::StrCat(
+            "CONNECT ", options_->http11ProxyInfo()->hostname,
+            options_->http11ProxyInfo()->hostname.rfind(':') == std::string::npos ? ":443" : "",
```
Comment (alyssawilk) [inline source/extensions/transport_sockets/http_11_proxy/connect.cc:43]: there should be utils somewhere for actual port parsing.  I'm not sure if v6 addresses are legal as proxy hosts but let's make sure we don't treat ::1 as having ports

**Review by alyssawilk — COMMENTED** [review]
thanks for the find and fix!
/wait

File: changelogs/current.yaml (hunk 1 above)
Comment (fredyw) [reply changelogs/current.yaml]: Done.

File: source/extensions/transport_sockets/http_11_proxy/connect.cc (hunk 2 above)
Comment (fredyw) [reply source/extensions/transport_sockets/http_11_proxy/connect.cc]: Done.

**Review by alyssawilk — APPROVED** [review]
LGTM modulo having an e2e test where we have a request original heading to host.com:1234 and a proxy config for proxy.com:6789 and we verify we handle the differing ports.  Totally fine if that test is in the hermetic mobile follow-up

**Comment by fredyw** [comment]
> LGTM modulo having an e2e test where we have a request original heading to host.com:1234 and a proxy config for proxy.com:6789 and we verify we handle the differing ports. Totally fine if that test is in the hermetic mobile follow-up

Yup, I have a separate PR specific to Envoy Mobile that will test this end-to-end.

**Comment by alyssawilk** [comment]
@phlax ideas about
hecking per-extension coverage
Per-extension coverage failed:
Code coverage for source/common/common/posix is lower than limit of 96.6 (96.2)

https://github.com/envoyproxy/envoy/blob/main/test/per_file_coverage.sh#L9 has this at 96.2 
I was experimenting with bumping it but that PR is not submitted

https://github.com/envoyproxy/envoy/pull/36075

**Comment by alyssawilk** [comment]
oh nm it was bumped down and needs a main merge

**Pull request merged**